olive/auto_optimizer/regulate_mixins.py

```diff
@@ -0,0 +1,131 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+from copy import deepcopy
+
+from olive.hardware.accelerator import Device
+
+
+class RegulatePassConfigMixin:
+    def regulate_pass_flows_dict(self, pass_flows_dict):
+        # remove useless passes according to the olive model type, for example if onnx model
+        # the conversion pass will be removed
+        if self.input_model_config.type.lower().endswith("onnxmodel"):
+            for pfs in pass_flows_dict.values():
+                for pf in pfs:
+                    pf.remove("OnnxConversion")
+
+        # special passes: OrtTransformerOptimization and OrtPerfTuning can be used for both fp16 and fp32
+        # we need assign different pass name for them
+        # for example: gpu_cuda_fp16, we need rename OrtTransformerOptimization to OrtTransformerOptimization_cuda_fp16
+        pass_flows_by_fp16 = pass_flows_dict.get("fp16", [])
+        pass_config, pass_flows_16 = self._regulate_fp16(None, pass_flows_by_fp16)
+
+        # flatten pass_flows_dict to pass_flows and generate the default pass_configs
+        pass_flows = []
+        unique_pass_flows = set()
+        if pass_flows_16:
+            pass_flows_dict["fp16"] = pass_flows_16
+        for pfs in pass_flows_dict.values():
+            for pf in pfs:
+                if tuple(pf) not in unique_pass_flows:
+                    pass_flows.append(pf)
+                unique_pass_flows.add(tuple(pf))
+                for p in pf:
+                    if p not in pass_config:
+                        pass_config.update({p: {"type": p, "config": {}}})
+
+        return pass_config, pass_flows
+
+    def _regulate_fp16(self, pass_config, pass_flows):
+        pass_config = pass_config or {}
+        is_gpu = self.accelerator_spec.accelerator_type == Device.GPU and self.accelerator_spec.execution_provider in [
```

Comment:
How are we going to support other gpu ep https://github.com/microsoft/Olive/blob/main/olive/hardware/accelerator.py#L56?

Reply:
1. For openvino and cpu ep in gpu device, we will ignore to regulate fp16.
2. For rocm, I am not sure how it works for fp16 which is converted from transformer optimization passes. If needed, we can just add it into the check lists.

Reply:
We add rocm use case when needed.